extensions/amp-instagram/0.1/amp-instagram.js

```diff
@@ -48,31 +52,110 @@ class AmpInstagram extends AMP.BaseElement {
   }
 
   /** @override */
-  isLayoutSupported(layout) {
-    return isLayoutSizeDefined(layout);
-  }
+  buildCallback() {
+    /**
+     * @private {?Element}
+     */
+    this.iframe_ = null;
+    /**
+     * @private {?Promise}
+     */
+    this.iframePromise_ = null;
 
-  /** @override */
-  layoutCallback() {
-    const width = this.element.getAttribute('width');
-    const height = this.element.getAttribute('height');
-    const shortcode = AMP.assert(
+    /**
+     * @private @const
+     */
+    this.shortcode_ = AMP.assert(
         (this.element.getAttribute('data-shortcode') ||
         this.element.getAttribute('shortcode')),
         'The data-shortcode attribute is required for <amp-instagram> %s',
         this.element);
-    // See
-    // https://instagram.com/developer/embedding/?hl=en
+  }
+
+  /** @override */
+  prerenderAllowed() {
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  maybeRenderIframe_() {
+    if (this.iframePromise_) {
+      return this.iframePromise_;
+    }
     const iframe = document.createElement('iframe');
+    this.iframe_ = iframe;
     iframe.setAttribute('frameborder', '0');
     iframe.setAttribute('allowtransparency', 'true');
     iframe.src = 'https://www.instagram.com/p/' +
-        encodeURIComponent(shortcode) + '/embed/?v=4';
+        encodeURIComponent(this.shortcode_) + '/embed/?v=4';
     this.applyFillContent(iframe);
-    iframe.width = width;
-    iframe.height = height;
+    iframe.width = this.element.getAttribute('width');
+    iframe.height = this.element.getAttribute('height');
     this.element.appendChild(iframe);
-    return loadPromise(iframe);
+    setStyles(iframe, {
+      'opacity': 0,
+    });
+    return this.iframePromise_ = loadPromise(iframe).then(() => {
+      this.getVsync().mutate(() => {
+        setStyles(iframe, {
+          'opacity': 1,
+        });
+      });
+    });
+  }
+
+  /** @override */
+  layoutCallback() {
+    const image = new Image();
+    // This will redirect to the image URL. By experimentation this is
+    // always the same URL that is actually used inside of the embed.
+    image.src = 'https://www.instagram.com/p/' +
```

Comment:
Actually this brings another question, since this won't actually prerender the image then if used on top view it'll be blank until we actually render no? 

Shouldn't we aim for prerendering it as a poster similar to what we do in `amp-youtube` and I believe `amp-img`?